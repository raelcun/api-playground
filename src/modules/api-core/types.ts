import * as t from 'io-ts'
import Koa from 'koa'

import { RolesV } from '@lib/rbac'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
export type KoaContext2<T> = Koa.ParameterizedContext<Koa.DefaultState, { request: { body: T } }>

type a = Koa.Middleware<Koa.DefaultState, { request: { body: T } }>

export type Middleware<T> = (ctx: KoaContext<T>, next: () => Promise<void>) => Promise<void>

export const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>
