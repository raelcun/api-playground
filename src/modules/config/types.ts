import * as t from 'io-ts'

import { LoggingConfig } from '@lib/logger'
import { DeepPartial, NonEmptyString } from '@lib/utils'

export interface FullConfig {
  isProduction: boolean
  env: string
  server: {
    jwtSecret: string
    securePort: number
    insecurePort: number
    enableSSL: boolean
  }
  logging: LoggingConfig
  aws: {
    region: string
    credentials: {
      accessKeyId: string
      secretAccessKey: string
    }
  }
}

export type PartialConfig<T> = DeepPartial<T>

export type ConfigProvider = () => FullConfig

export const envVarsV = t.type({
  NODE_ENV: NonEmptyString,
  JWT_SECRET: NonEmptyString,
  AWS_ACCESS_KEY_ID: NonEmptyString,
  AWS_SECRET_ACCESS_KEY: NonEmptyString,
})
export type EnvVars = t.TypeOf<typeof envVarsV>
