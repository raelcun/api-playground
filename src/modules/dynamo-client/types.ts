import { DocumentClient } from 'aws-sdk/clients/dynamodb'
import { AWSError } from 'aws-sdk/lib/error'
import { Response } from 'aws-sdk/lib/response'
import { taskEither as TE } from 'fp-ts'

import { Err } from '@modules/error/types'
import { ChangeTypeOfKeys } from '@root/utils/types'

export interface DynamoClient<T> {
  put: (
    params: ChangeTypeOfKeys<DocumentClient.PutItemInput, 'Item', T>,
  ) => TE.TaskEither<Err, Response<DocumentClient.PutItemOutput, AWSError>>
  delete: (
    params: DocumentClient.DeleteItemInput,
  ) => TE.TaskEither<Err, Response<DocumentClient.DeleteItemOutput, AWSError>>
}
