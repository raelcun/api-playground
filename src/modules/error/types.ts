export interface Err<Code extends string = string> {
  code: Code
  message?: string
}
