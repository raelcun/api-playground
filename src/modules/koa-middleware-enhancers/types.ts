import Koa from 'koa'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
