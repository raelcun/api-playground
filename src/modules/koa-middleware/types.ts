import { KoaContext } from '../koa-middleware-enhancers'

export type Middleware<T> = (
  ctx: KoaContext<T>,
  next: () => Promise<void>,
) => Promise<void>
