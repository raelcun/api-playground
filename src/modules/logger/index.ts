import pino from 'pino'
import { getConfig } from '../../config'
import { Logger } from './types'

export const getLogger = (name: string) => {
  return pino({ name, level: getConfig().logging.level })
}

export const getSystemLogger = (): Logger => getLogger('system')

export const createMockLogger = (
  overrides: Partial<
    Pick<pino.Logger, 'fatal' | 'trace' | 'error' | 'info' | 'debug' | 'warn' | 'child'>
  > = {},
): Logger => {
  return Object.assign(pino({ name: 'mockLogger', level: 'trace' }), overrides)
}
