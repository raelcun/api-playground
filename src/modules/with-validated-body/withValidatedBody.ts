import t from 'io-ts'
import HttpStatus from 'http-status-codes'
import { reporter } from 'io-ts-reporters'
import { left } from 'fp-ts/lib/Either'
import { getSystemLogger } from '../logger'
import { Middleware } from '../../types'

export const withValidatedBody = <T>(type: t.Type<T, unknown>) => (
  middleware: Middleware<T>,
): Middleware<T> => async (ctx, next) => {
  await type.decode(ctx.request.body).fold(
    async errors => {
      ctx.status = HttpStatus.BAD_REQUEST
      ctx.body = {
        errors: reporter(left(errors)),
      }
      getSystemLogger().trace(ctx.body, `body validation failed for ${ctx.url}`)
    },
    async validatedBody => {
      ctx.request.body = validatedBody
      await middleware(ctx, next)
    },
  )
}
