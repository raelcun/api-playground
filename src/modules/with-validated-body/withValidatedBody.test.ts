import * as t from 'io-ts'
import { createMockContext } from '@shopify/jest-koa-mocks'
import HttpStatus from 'http-status-codes'
import { Middleware } from '../../types'
import { createMockLogger } from '../logger'
import { withValidatedBodyInner } from './withValidatedBody'

describe('validateRequestBody', () => {
  describe('when validation is successful', () => {
    test('should invoke wrapped middleware', async () => {
      expect.assertions(1)

      const wrappedMiddleware: Middleware<number> = async ctx => {
        expect(typeof ctx.request.body).toEqual('number')
      }

      const middleware = withValidatedBodyInner(createMockLogger())(t.number)(wrappedMiddleware)

      await middleware(createMockContext({ requestBody: 5 }), jest.fn())
    })
  })

  describe('when validation is unsuccessful', () => {
    const wrappedMiddleware = jest.fn()
    const context = createMockContext({ requestBody: '5' })
    const loggerOverrides = { trace: jest.fn() }

    beforeAll(async () => {
      const middleware = withValidatedBodyInner(createMockLogger(loggerOverrides))(t.number)(
        wrappedMiddleware,
      )

      await middleware(context, jest.fn())
    })

    test('should not invoke wrapped middleware', () => {
      expect(wrappedMiddleware).not.toHaveBeenCalled()
    })

    test('should have a status code of 400', () => {
      expect(context.status).toEqual(HttpStatus.BAD_REQUEST)
    })

    test('should have a body with error message', () => {
      expect(context.body).toMatchSnapshot()
    })

    test('should have logged validation error', () => {
      expect(loggerOverrides.trace.mock.calls[0][0]).toMatchSnapshot()
      expect(loggerOverrides.trace.mock.calls[0][1]).toMatchSnapshot()
    })
  })
})
