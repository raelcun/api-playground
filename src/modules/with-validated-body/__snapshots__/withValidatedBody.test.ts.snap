// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRequestBody when validation is unsuccessful should have a body with error message 1`] = `
Object {
  "errors": Array [
    "Expecting number but instead got: \\"5\\".",
  ],
}
`;

exports[`validateRequestBody when validation is unsuccessful should have logged validation error 1`] = `
Object {
  "requestBody": "5",
  "validationError": Array [
    "Expecting number but instead got: \\"5\\".",
  ],
}
`;

exports[`validateRequestBody when validation is unsuccessful should have logged validation error 2`] = `"body validation failed for http://test.com/"`;
